Add Cargo command for cross-compiling and packaging buildpacks (#199)
Malax/libcnb.rs: README.md

    // more information about platforms:
    // https://github.com/buildpacks/spec/blob/main/buildpack.md
    type Platform = GenericPlatform;

    // The type for the metadata of the buildpack itself. This is the data found in the
    // `[metadata]` section of your buildpack's `buildpack.toml`. The framework will automatically
    // try to parse it into the specified type. This example buildpack uses GenericMetadata which
    // provides low-level access to the TOML table.
    type Metadata = GenericMetadata;

    // The error type for this buildpack. Buildpack authors usually implement an enum with
    // specific errors that can happen during buildpack execution. This error type should
    // only contain error specific to this buildpack, such as `CouldNotExecuteMaven` or
    // `InvalidGemfileLock`. This example buildpack uses `GenericError` which means this buildpack
    // does not specify any errors.
    //
    // More generic errors that happen during buildpack execution such as I/O errors while
    // writing CNB TOML files are handled by libcnb.rs itself.
    type Error = GenericError;

>     // This method will be called when the CNB lifecycle calls detect. Use the `DetectContext` to

Author: Can you open a dedicated PR for this if you feel strongly? This comment wasn't added in this PR and I'd like to keep this as clean as possible.

>     // This method will be called when the CNB lifecycle calls detect. Use the `DetectContext` to

Reviewer: https://github.com/Malax/libcnb.rs/pull/216

    // access CNB data such as the stack this buildpack is currently executed on, the app
    // directory and similar things. When using libcnb.rs, you never have to read environment
    // variables or read/write files to disk to interact with the CNB lifecycle.
    //
    // One example of this is the return type of this method. `DetectResult` encapsulates the
    // required exit code as well as the data written to the build plan. libcnb.rs will,
    // according to the returned value, handle both writing the build plan and exiting with
    // the correct status code for you.
    fn detect(&self, _context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
        DetectResultBuilder::pass().build()
    }

    // Similar to detect, this method will be called when the CNB lifecycle executes the

>     // build phase.

Author: "build" is the proper name of the phase: https://github.com/buildpacks/spec/blob/74f91666fc22e7568acdd0ca9e2e314ff150d740/buildpack.md#phase-3-build

    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
        println!("Hello World!");
        println!("Build runs on stack {}!", context.stack_id);

        BuildResultBuilder::new()
            .launch(Launch::new().process(Process::new(
                process_type!("web"),
                "echo",
                vec!["Hello World!"],
                false,

>                 true,

Author: This is something for `Process` itself, and IMO out of scope for this PR. Do you want to file an issue for improving `Process::new`?

            )))
            .build()
    }
}

// Implements the main function and wires up the framework for the given buildpack.
buildpack_main!(HelloWorldBuildpack);
```

### Packaging the Buildpack
Now that our buildpack is written, it's time to package it, so that it can be run. If you followed the steps to setup
your development environment, you have access to the `libcnb` Cargo command that will handle packaging for you.

In your project directory, run `cargo libcnb package` to start packaging:

```shell
$ cargo libcnb package
INFO - Reading buildpack metadata...
INFO - Found buildpack libcnb-examples/my-buildpack with version 0.1.0.
INFO - Building buildpack binary (x86_64-unknown-linux-musl)...
   Compiling syn v1.0.82
   Compiling bit-vec v0.6.3
   Compiling regex-syntax v0.6.25
   # Omitting further compilation output...
    Finished dev [unoptimized + debuginfo] target(s) in 19.71s
INFO - Writing buildpack directory...
INFO - Successfully wrote buildpack directory: ../../target/libcnb-examples_ruby_0.1.0_dev (103.6M)
INFO - Packaging successfully finished!

> INFO - Hint: To test your buildpack locally with pack, run: pack build my-image --buildpack ../../target/libcnb-examples_ruby_0.1.0_dev --path /path/to/application

Author: This is caused by incorporating suggestions from review comments and not updating the example output. I'll update them after the dust settles on this PR.

```

If you get errors with hints about how to install required tools to cross-compile from your host platform to the
target platform, follow them and try again. These differ from platform to platform, which is why they're not part of this
quick start guide.

### Running the Buildpack
You might have seen in the hint in the output of the packaging step that we're now ready to run our buildpack locally 
with `pack`. Before we can do this, we need an application to build. Since our buildpack does not interact with the
application code at all, we just create an empty directory and use that as our application:

```shell
$ mkdir bogus-app

> $ pack build my-image -b target/libcnb-examples_my-buildpack_0.1.0_dev.tar.gz --path bogus-app --builder heroku/buildpacks:20

Author: Yes, we don't know which default builder the user has set in `pack`, but this buildpack requires `heroku-20`.

It currently targets version `0.6` of the CNB spec.

## Quick Start Guide
This quick start guide walks you through writing a simple Cloud Native Buildpack in Rust with libcnb.rs. It will also 
demonstrate how to package your buildpack and how to build an application image with it.

### Development Environment Setup
In addition to the libcnb crate, we need some tools to compile, package and run the buildpack. These steps only need to
be carried out once and don't need to be repeated for each buildpack you will write. 

> #### libcnb Cargo Command

Reviewer 2: ```suggestion

#### libcnb Cargo Command
```

the use of buildpacks. It's the tool that we will eventually use to run our buildpack and create application images. 
Find their documentation about installing it here: https://buildpacks.io/docs/tools/pack/

### Project Setup

After we've successfully set up our development environment, we can move on and create a new Rust project for our 
buildpack. First, we create a new binary crate with `Cargo`:

```shell
$ cargo new my-buildpack
     Created binary (application) `my-buildpack` package
```

Then, modify your project's `Cargo.toml` to include the `libcnb` dependency:

```toml
[package]
name = "my-buildpack"
version = "0.1.0"
edition = "2021"

[dependencies]
libcnb = "0.4.0"
```

Since we're writing a Cloud Native Buildpack, we also need a `buildpack.toml`. Use the template below and write it to a
file named `buildpack.toml` in the root of your project, right next to `Cargo.toml`.

```toml
api = "0.6"

[buildpack]
id = "libcnb-examples/my-buildpack"
version = "0.1.0"
name = "My Buildpack"

[[stacks]]
id = "heroku-20"
```

That's all we need! We can now move on to finally write some buildpack code!

### Writing the Buildpack

> As aforementioned, the buildpack we're writing will be very simple. We will just log a "Hello World" image during build

Reviewer 2: ```suggestion
As aforementioned, the buildpack we're writing will be very simple. We will just log a "Hello World" message during the build
```

and set the default process type to a command that will also emit "Hello World" when the application image is run. 
Find more complex example buildpacks in the [examples directory](examples).

Modify the project's `src/main.rs` file to contain the following:

```rust,no_run
use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
use libcnb::data::launch::{Launch, Process};
use libcnb::data::process_type;
use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
use libcnb::generic::{GenericError, GenericMetadata, GenericPlatform};
use libcnb::{buildpack_main, Buildpack};

struct HelloWorldBuildpack;

impl Buildpack for HelloWorldBuildpack {

>     // The CNB platform this buildpack targets, usually GenericPlatform. See the CNB spec for

Reviewer 2: ```suggestion
    // The CNB platform this buildpack targets, usually `GenericPlatform`. See the CNB spec for
```

    // more information about platforms:
    // https://github.com/buildpacks/spec/blob/main/buildpack.md
    type Platform = GenericPlatform;

    // The type for the metadata of the buildpack itself. This is the data found in the

>     // `[metadata]` section of your buildpack's buildpack.toml. The framework will automatically

Reviewer 2: ```suggestion
    // `[metadata]` section of your buildpack's `buildpack.toml`. The framework will automatically
```

    // try to parse it into the specified type. This example buildpack uses GenericMetadata which
    // provides low-level access to the TOML table.
    type Metadata = GenericMetadata;

    // The error type for this buildpack. Buildpack authors usually implement an enum with
    // specific errors that can happen during buildpack execution. This error type should

>     // only contain error specific to this buildpack, such as CouldNotExecuteMaven or
>     // InvalidGemfileLock. This example buildpack uses GenericError which means this buildpack

Reviewer 2: ```suggestion
    // only contain error specific to this buildpack, such as `CouldNotExecuteMaven` or
    // `InvalidGemfileLock`. This example buildpack uses `GenericError` which means this buildpack
```

    // does not specify any errors.
    //
    // More generic errors that happen during buildpack execution such as I/O errors while
    // writing CNB TOML files are handled by libcnb.rs itself.
    type Error = GenericError;

>     // This method will be called when the CNB lifecycle calls detect. Use the DetectContext to

Reviewer 2: ```suggestion
    // This method will be called when the CNB lifecycle calls detect. Use the `DetectContext` to
```

    // access CNB data such as the stack this buildpack is currently executed on, the app
    // directory and similar things. When using libcnb.rs, you never have to read environment
    // variables or read/write files to disk to interact with the CNB lifecycle.
    //

>     // One example of this is the return type of this method. DetectResult encapsulates the

Reviewer 2: ```suggestion
    // One example of this is the return type of this method. `DetectResult` encapsulates the
```

    // required exit code as well as the data written to the build plan. libcnb.rs will,
    // according to the returned value, handle both writing the build plan and exiting with
    // the correct status code for you.
    fn detect(&self, _context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
        Ok(DetectResultBuilder::pass().build())
    }

    // Similar to detect, this method will be called when the CNB lifecycle executes the
    // build phase.
    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
        println!("Hello World!");
        println!("Build runs on stack {}!", context.stack_id);

        Ok(BuildResultBuilder::new()
            .launch(Launch::new().process(Process::new(
                process_type!("web"),
                "echo",
                vec!["Hello World!"],
                false,
                true,
            )))
            .build())
    }
}

// Implements the main function and wires up the framework for the given buildpack.
buildpack_main!(HelloWorldBuildpack);

> 

Reviewer 2: (Means when using the "copy" button in the rendered markdown, the resultant text doesn't have two trailing empty lines)

```suggestion
```

### Packaging the Buildpack
Now that our buildpack is written, it's time to package it, so that it can be run. If you followed the steps to setup
your development environment, you have access to the `libcnb` Cargo command that will handle packaging for you.

In your project directory, run `cargo libcnb package` to start packaging:

```shell
$ cargo libcnb package
INFO - Reading buildpack metadata...
INFO - Found valid buildpack with id "libcnb-examples/my-buildpack" @ 0.1.0!
INFO - Building buildpack binary (x86_64-unknown-linux-musl)...
   Compiling syn v1.0.82
   Compiling bit-vec v0.6.3
   Compiling regex-syntax v0.6.25
   # Omitting further compilation output...
    Finished dev [unoptimized + debuginfo] target(s) in 19.71s
INFO - Writing buildpack tarball...
INFO - Successfully wrote buildpack tarball target/libcnb-examples_my-buildpack_0.1.0_dev.tar.gz (6.5M)
INFO - Packaging successfully finished!
INFO - Hint: To test your buildpack locally with pack, run: pack build my-image -b target/libcnb-examples_my-buildpack_0.1.0_dev.tar.gz --path /path/to/application
```

If you get errors with hints about how to install required tools to cross-compile from your host platform to the

> target platform, follow them and try again. These differ from platform to platform, that's why they're not part of this

Reviewer 2: ```suggestion
target platform, follow them and try again. These differ from platform to platform, which is why they're not part of this
```

quick start guide.

### Running the Buildpack
You might have seen in the hint in the output of the packaging step that we're now ready to run our buildpack locally 
with `pack`. Before we can do this, we need an application to build. Since our buildpack does not interact with the
application code at all, we just create an empty directory and use that as our application:

```shell
$ mkdir bogus-app

> $ pack build my-image -b target/libcnb-examples_my-buildpack_0.1.0_dev.tar.gz --path bogus-app

Reviewer 2: Will this fail if the user's pack config still has the initial default builder? (Which I think might be the samples builder, I can't remember)

If so, we either need to pass `--builder heroku/buildpacks:20` here, or have making it the default (https://buildpacks.io/docs/tools/pack/cli/pack_config_default-builder/) be part of the initial setup section earlier.

Author: Good catch, I didn't try. I'll figure this out and update the docs accordingly.

Author: I added an explicit builder flag in the guide. Again, great catch!

20: Pulling from heroku/buildpacks
Digest: sha256:04e8ea7a1f482f289d432d9518edcfaaf9f3a10432cd1b624e58225f22e7c416
Status: Image is up to date for heroku/buildpacks:20
20: Pulling from heroku/pack
Digest: sha256:21ea4b85bc937b47e017bf43136a5295ec08e093e3c210b20e69e9c1b0f2bd57
Status: Image is up to date for heroku/pack:20
===> DETECTING
libcnb-examples/my-buildpack 0.1.0
===> ANALYZING
Skipping buildpack layer analysis
===> BUILDING
Hello World!
Build runs on stack heroku-20!
===> EXPORTING
Reusing 1/1 app layer(s)
Reusing layer 'launcher'
Reusing layer 'config'
Reusing layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving my-image...
*** Images (d4f67a828236):
      my-image
Successfully built image my-image
```

### Running the image

> Just run the crated Docker image as you normally would. If all went well, you should see our "Hello World!" message in
> your terminal:

Reviewer 2: "Just" can sometimes read a bit negative ("oh you just have to do X, it's not that hard"). The usage here isn't too bad, but perhaps something like this would be gentler?

```suggestion
The newly created Docker image can be run in the same way as you would a Docker image created via `docker build`.
If all went well, you should see our "Hello World!" message in your terminal:
```

```shell
$ docker run my-image
Hello World!
```

### Next Steps
While the buildpack we've written in this quick start guide is not very useful, it can serve as a starting point for a 
more useful buildpack. To discover more of the libcnb API, browse the [examples directory](example) and the 

> [documentation on docs.rs](docs.rs).

Reviewer 2: The link is currently broken, I think this will fix it, but haven't checked.

```suggestion
[documentation on docs.rs][docs.rs].
```

Author: This should be a direct link to the libcnb.rs docs. I'll fix it! :)

Reviewer 2: There is already an named link of `[libcnb.rs]` used at the top of the doc

Author: Oooooh, I see now. That's clever.
